sat/cli/bootsys/platform.py

```diff
@@ -377,6 +386,54 @@ def run(self):
             self._success_exit(f'All containers stopped on {self.host}.')
 
 
+class FilteredHostKeys(HostKeys):
+    """
+    Host keys loader which filters all hosts except the specified ones.
+
+    This is needed to speed up known_hosts file parsing as paramiko's implementation
+    (up to and including 3.2.0) has quadratic time complexity. To get around this,
+    we can hackily cut down a large known hosts file to just a few hosts in linear
+    time before parsing, which should help tremendously in the case of a very large
+    known_hosts file.
+    """
+    def __init__(self, filename=None, hostnames=None):
+        """Construct a FilteredHostKeys object.
+
+        Args:
+            filename (str): path to the known hosts file
+            hostnames ([str]): list of hostnames for which to load host keys from the
+                known_hosts file
+        """
+        self.hostnames = hostnames
+        super().__init__(filename)
+
+    def load(self, filename):
+        """Load known hosts from the specified known_hosts file.
+
+        Args:
+            filename (str): path to the known_hosts file
+        """
+        if not self.hostnames:
+            return super().load(filename)
+
+        with open(filename) as known_hosts, \
+                tempfile.NamedTemporaryFile(mode='w+') as tmp_known_hosts:
+            for line in known_hosts:
+                entry = HostKeyEntry.from_line(line)
+                if any(hn in entry.hostnames for hn in self.hostnames):
+                    tmp_known_hosts.write(entry.to_line())
+            tmp_known_hosts.flush()
+            super().load(tmp_known_hosts.name)
+
+    def save(self, filename):
+        """(Don't) save the known_hosts file.
+
+        This is a no-op since we don't want to accidentally destructively
+        modify the user's known hosts file.
+        """
+        return NotImplemented
```

Comment:
Did you perhaps mean to `raise NotImplementedError`? Or perhaps this should just be an error log message?

Reply:
I actually thought before that `NotImplemented` was just a singleton without special properties, akin to `None` or `Ellipsis`, but apparently `return NotImplemented` does have special meaning. According to [the docs](https://docs.python.org/3/library/constants.html#NotImplemented), if it's returned by a special binary operator, e.g. `__eq__()` or `__le__()`, then CPython will try the other type's corresponding operation (e.g. `__eq__()` reflexively or `__gt__()`), so for non magic methods it can cause weird error messages. I'll just change it to return `None`, I don't want to raise exceptions in a method other library code can use.

Reply:
Yeah, that seems reasonable. I'm not expecting we will actually save the hosts file anywhere, but I haven't dug through paramiko to see how this class is used. I wonder if we could also log a warning or something.